atakama/rule_engine.py

```diff
@@ -109,6 +115,40 @@ def from_dict(cls, data: dict) -> "RulePlugin":
         assert isinstance(p, RulePlugin), "Rule plugins must derive from RulePlugin"
         return p
 
+    def to_dict(self):
+        out = self.args.copy()
+        out["rule"] = self.name()
+        return out
+
+
+class RuleIdGenerator:
+    """Manage unique rule id generation."""
+
+    __autodoc__ = False
+
+    def __init__(self):
+        self._seen = defaultdict(lambda: 0)
+        self._rule_seq = 0
+
+    def inject_rule_id(self, ent: dict):
+        """
+        Modify the supplied dictionary to add a rule_id, but only if a rule_id is not present.
+
+        Keeps track of rule id's and ensures uniqueness, while trying to maintain consistency.
+        """
+
+        self._rule_seq += 1
+        rule_id = ent.get("rule_id")
+        if not rule_id:
+            ent_data = json.dumps(ent, sort_keys=True, separators=(",", ":"))
+            ent_hash = hashlib.md5(ent_data.encode("utf8")).hexdigest()
+            # if the hash is enough, use it, that way it's relocatable and still consistent
+            if ent_hash in self._seen:
+                # otherwise append a sequence
+                ent_hash += "." + str(self._rule_seq)
+            rule_id = ent["rule_id"] = ent_hash
+        self._seen[rule_id] += 1
```

Comment:
are we meant to throw somewhere if this is > 1?

Reply:
no